src/mlpack/core/tree/spill_tree/spill_tree_impl.hpp

```diff
@@ -0,0 +1,731 @@
+{
+  if (!Parent())
+    return true;
+
+  const bool left = this == Parent()->Left();
+
+  if (left)
+    return !Parent()->Hyperplane().Right(other.Bound());
+  else
+    return !Parent()->Hyperplane().Left(other.Bound());
+}
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+template<typename VecType>
+inline bool SpillTree<MetricType, StatisticType, MatType, HyperplaneType,
+    SplitType>::HalfSpaceContains(
+    const VecType& point,
+    typename boost::enable_if<IsVector<VecType> >::type*) const
+{
+  if (!Parent())
+    return true;
+
+  const bool left = this == Parent()->Left();
+  const bool toTheLeft = Parent()->Hyperplane().Left(point);
+
+  return left == toTheLeft;
+}
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitMetricType, typename SplitMatType>
+             class SplitType>
+void SpillTree<MetricType, StatisticType, MatType, HyperplaneType, SplitType>::
+    SplitNode(std::vector<size_t>& points,
+              const size_t maxLeafSize,
+              const double tau,
+              const double rho)
+{
+  // We need to expand the bounds of this node properly.
+  for (size_t i = 0; i < points.size(); i++)
+    bound |= dataset->cols(points[i], points[i]);
+
+  // Calculate the furthest descendant distance.
+  furthestDescendantDistance = 0.5 * bound.Diameter();
+
+  // Now, check if we need to split at all.
+  if (points.size() <= maxLeafSize)
+  {
+    pointsIndex = new std::vector<size_t>();
+    pointsIndex->swap(points);
+    count = pointsIndex->size();
+    return; // We can't split this.
+  }
+
+  const bool split = SplitType<MetricType, MatType>::SplitSpace(bound,
+      *dataset, points, hyperplane);
+  // The node may not be always split. For instance, if all the points are the
+  // same, we can't split them.
+  if (!split)
+  {
+    pointsIndex = new std::vector<size_t>();
+    pointsIndex->swap(points);
+    count = pointsIndex->size();
+    return; // We can't split this.
+  }
+
+  std::vector<size_t> leftPoints, rightPoints;
+  // Split the node.
+  overlappingNode = SplitPoints(tau, rho, points, leftPoints, rightPoints);
+
+  // We don't need the information in points, so lets clean it.
+  std::vector<size_t>().swap(points);
+
+  // Now we will recursively split the children by calling their constructors
+  // (which perform this splitting process).
+  left = new SpillTree(this, leftPoints, tau, maxLeafSize, rho);
+  right = new SpillTree(this, rightPoints, tau, maxLeafSize, rho);
+
+  // Update count number, to represent the number of descendant points.
+  count = left->NumDescendants() + right->NumDescendants();
```

Comment:
I think there is an easier way to do this.  We only hold a list of point indices in the leaves, so we only have to worry about the arrangement of points at the leaf level.  Each leaf node only has one sibling, so given two sibling leaf nodes N_l and N_r, let's define a couple sets.  I've just chosen arbitrary set names:

 - S: all the points held in N_l and N_r
 - T: points held in N_r but _not_ in N_l
 - U: points held in both N_l and N_r
 - V: points held in N_l

Now we can arrange the point sets for each child like this:

 - N_l: [V] (order does not matter here)
 - N_r: [T U] (order does matter: all points in T come before all points in U)

If we have set the number of descendants in the parent properly to be |S| (this should be easy), then we can do the following:

```
size_t Descendant(const size_t i)
{
  // Are we a leaf?
  if (IsLeaf())
    return (*pointsIndex)[index];

  if (i > left->NumDescendants())
    return left->Descendant(i);
  else
    return right->Descendant(i - left->NumDescendants());
}
```

The key to this approach is determining the sets T and U, but I think that is straightforward to do either before you recurse into building the children (you can "pre-organize" the set you pass to the right child), or afterwards (you can access the sets of the left and right child and organize them).

When you set `numDescendants` when building a node, that would simply be equal to the number of points held in that node.  This means for a given node, it is possible (and okay) if `node->NumDescendants() < node->Left()->NumDescendants() + node->Right()->NumDescendants()`.

Let me know if I have overlooked anything here.